File: web/lib/domain/difficulty.ts
export interface DifficultyInput {
  distanceKm: number;
  ascentM: number;
  descentM: number;
}

export type DifficultyClass = 'easy' | 'moderate' | 'hard' | 'extreme';

export interface DifficultyResult {
  score: number;
  klass: DifficultyClass;
  effortKm: number;
}

// 100 m of climb ≈ 0.85 effort-km; descent adds lighter fatigue (0.25).
// Extreme threshold: 45 effort-km maps to score 100.
const ASCENT_W = 0.85;
const DESCENT_W = 0.25;
const EXTREME_EFFORT_KM = 45;

export function scoreDifficulty(i: DifficultyInput): DifficultyResult {
  const effortKm =
    i.distanceKm +
    (i.ascentM / 100) * ASCENT_W +
    (i.descentM / 100) * DESCENT_W;

  const score = Math.max(
    0,
    Math.min(100, Math.round((effortKm / EXTREME_EFFORT_KM) * 100)),
  );

  const klass: DifficultyClass =
    score <= 25 ? 'easy'
    : score <= 50 ? 'moderate'
    : score <= 75 ? 'hard'
    : 'extreme';

  return { score, klass, effortKm };
}

File: ios/WaypointiOS/WaypointiOS/Domain/Difficulty.swift
//
//  Difficulty.swift
//  WaypointiOS
//
//  Verbatim port of web/lib/domain/difficulty.ts.
//  Constants are canonical in ARCHITECTURE.md §10.1 — change there first.
//

import Foundation

// MARK: - Types

enum DifficultyClass: String, Codable, Sendable {
    case easy, moderate, hard, extreme
}

struct DifficultyInput: Sendable {
    let distanceKm: Double
    let ascentM: Double
    let descentM: Double
}

struct DifficultyResult: Sendable {
    let score: Int
    let klass: DifficultyClass
    let effortKm: Double
}

// MARK: - Constants (ARCHITECTURE.md §10.1)

private let ascentW: Double = 0.85
private let descentW: Double = 0.25
private let extremeEffortKm: Double = 45

// MARK: - Engine

func scoreDifficulty(_ i: DifficultyInput) -> DifficultyResult {
    let effortKm = i.distanceKm
        + (i.ascentM / 100) * ascentW
        + (i.descentM / 100) * descentW

    let score = max(0, min(100, Int((effortKm / extremeEffortKm * 100).rounded())))

    let klass: DifficultyClass =
        score <= 25 ? .easy
        : score <= 50 ? .moderate
        : score <= 75 ? .hard
        : .extreme

    return DifficultyResult(score: score, klass: klass, effortKm: effortKm)
}
